### PR TITLE
Implement fork.Factory

### PR DESCRIPTION
**Pull request by dstdfx**
## Which problem is this PR solving?
- Required changes for https://github.com/jaegertracing/jaeger/pull/2496

## Short description of the changes
- Implement fork.Factory that delegates metrics with
specific namespaces to different factory, otherwise
default factory is used.

### CODE REVIEW

**Comment by dstdfx** [comment]
cc @yurishkuro

File: metrics/fork/fork_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	defaultFactory := metricstest.NewFactory(time.Second)
+
+	// Create factory that will delegate namespaced metrics to forkFactory
+	// and add some metrics
+	ff := New(forkNamespace, forkFactory, defaultFactory)
+	ff.Gauge(metrics.Options{
+		Name: "somegauge",
+		Tags: nil,
+		Help: "",
+	}).Update(666)
```
Comment (yurishkuro) [inline metrics/fork/fork_test.go:44]: let's go with a less controversial number :-)

```suggestion
	}).Update(42)
```

File: metrics/fork/fork_test.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	forkFactory := metricstest.NewFactory(time.Second)
+	defaultFactory := metricstest.NewFactory(time.Second)
+
+	// Create factory that will delegate namespaced metrics to forkFactory
+	// and add some metrics
+	ff := New(forkNamespace, forkFactory, defaultFactory)
+	ff.Gauge(metrics.Options{
+		Name: "somegauge",
+		Tags: nil,
+		Help: "",
```
Comment (yurishkuro) [inline metrics/fork/fork_test.go:43]: no need to assign default values like nil and ""

File: Gopkg.lock [hunk 3]
```diff
@@ -196,6 +212,13 @@
     "github.com/stretchr/testify/assert",
     "github.com/stretchr/testify/require",
     "github.com/uber-go/tally",
+    "github.com/uber/jaeger-lib/metrics",
+    "github.com/uber/jaeger-lib/metrics/adapters",
+    "github.com/uber/jaeger-lib/metrics/expvar",
+    "github.com/uber/jaeger-lib/metrics/go-kit",
+    "github.com/uber/jaeger-lib/metrics/go-kit/expvar",
+    "github.com/uber/jaeger-lib/metrics/metricstest",
+    "github.com/uber/jaeger-lib/metrics/prometheus",
```
Comment (yurishkuro) [inline Gopkg.lock:221]: which version of `dep` are you using? Ideally there should be no dependency changes in this PR.

**Review by yurishkuro — COMMENTED** [review]
lgtm, a few small comments

File: Gopkg.lock (hunk 3 above)
Comment (dstdfx) [reply Gopkg.lock:221]:
```
dep:
 version     : v0.5.4
 build date  : 2019-09-29
 git hash    : 1f7c19e
 go version  : go1.13
 go compiler : gc
 platform    : darwin/amd64
 features    : ImportDuringSolve=false
```

File: metrics/fork/fork_test.go (hunk 2 above)
Comment (dstdfx) [reply metrics/fork/fork_test.go]: > no need to assign default values like nil and ""

Just realised that Goland-autocomplete does it for me sometimes...

File: metrics/fork/fork_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/uber/jaeger-lib/metrics"
+	"github.com/uber/jaeger-lib/metrics/metricstest"
+)
+
+func TestAPICheck(t *testing.T) {
+	var v interface{} = &Factory{}
+	_, ok := v.(metrics.Factory)
+	assert.True(t, ok)
+}
```
Comment (yurishkuro) [inline metrics/fork/fork_test.go:30]: Much simpler with a compile-time check:

```suggestion
var _ metrics.Factory = (*Factory)(nil)
```

**Review by yurishkuro — APPROVED** [review]

**Comment by yurishkuro** [comment]
👍

**Pull request merged**